File: Graphs/Medium/Topological Sort/topologicalsort.cpp
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

void util(int v, vector<vector<int>>& graph, vector<bool>& visited, stack<int>& st) {
    visited[v] = true;
    for (int i : graph[v]) {
        if (!visited[i]) {
            util(i, graph, visited, st);
        }
    }
    st.push(v); // Add the current vertex to the stack
}

void topologicalsort(vector<vector<int>>& graph) {
    int n = graph.size();
    vector<bool> visited(n, false);
    stack<int> st;
    for (int i = 0; i < n; i++) {
        if (!visited[i]) {
            util(i, graph, visited, st);
        }
    }
    while (!st.empty()) {
        cout << st.top() << " "; // Print the top element of the stack
        st.pop(); // Remove the top element from the stack
    }
    cout << endl;
}

int main() {
    int vertices;
    cout << "Enter the number of vertices: ";
    cin >> vertices;

    vector<vector<int>> graph(vertices);

    for (int i = 0; i < vertices; i++) {
        int num_neighbors;
        cout << "Enter the number of neighbors for vertex " << i << ": ";
        cin >> num_neighbors;

        graph[i].resize(num_neighbors);

        cout << "Enter neighbors for vertex " << i << ": ";
        for (int j = 0; j < num_neighbors; j++) {
            cin >> graph[i][j];
        }
    }

    topologicalsort(graph);

    return 0;
}

File: Graphs/Medium/Topological Sort/topologicalSort.py
def util(v, graph, visited, stack):
    visited[v] = True 
    for i in graph[v]:
        if not visited[i]:
            util(i, graph, visited, stack)
    stack.append(v)

def topological_sort(graph):
    visited = [False] * len(graph)
    stack = []

    for i in range(len(graph)):
        if not visited[i]:
            util(i, graph, visited, stack)

    print(stack[::-1])

if __name__ == "__main__":
    graph = {}

    print("Enter the number of vertices: ")
    vertices = int(input())

    for i in range(vertices):
        print(f"Enter the neighbors of vertex {i} separated by space (or press Enter for no neighbors):")
        neighbors = list(map(int, input().split()))
        graph[i] = neighbors

    topological_sort(graph)